File: 백준/Gold/1633. 최고의 팀 만들기/최고의 팀 만들기.py
import sys

arr = sys.stdin.readlines()
arr = [list(map(int, a.strip().split())) for a in arr]
cache = [[[0 for _ in range(16)] for _ in range(16)] for _ in range(len(arr))]
cache[0][1][0] = arr[0][0]
cache[0][0][1] = arr[0][1]

for i in range(1, len(arr)):
    for j in range(16):
        for k in range(16):
            cache[i][j][k] = cache[i - 1][j][k]
            if j > 0:
                cache[i][j][k] = max(cache[i][j][k], cache[i - 1][j - 1][k] + arr[i][0])
            if k > 0:
                cache[i][j][k] = max(cache[i][j][k], cache[i - 1][j][k - 1] + arr[i][1])


print(cache[len(arr) - 1][15][15])

File: 백준/Gold/1633. 최고의 팀 만들기/최고의 팀 만들기.cc
#include <bits/stdc++.h>
using namespace  std;
int cache[1005][16][16];
vector<array<int, 2>> arr;
int n;
int solve(int idx, int w, int b){
    if (idx == n)
        return (w == 15 && b == 15 ? 0 : -987654321);
    int &ret = cache[idx][w][b];
    if(ret != -1) return ret;
    ret = 0;
    ret = solve(idx + 1, w, b);
    if (w < 15)
        ret = max(ret, solve(idx + 1, w + 1, b) + arr[idx][0]);
    if (b < 15)
        ret = max(ret, solve(idx + 1, w, b + 1) + arr[idx][1]);
    return ret;
}

int main(){
    ios_base::sync_with_stdio(0); cin.tie(0);
    memset(cache, -1 ,sizeof(cache));
    
    int a, b;
    while (cin >> a >> b)
    {
        arr.push_back({a, b});
    }
    n = arr.size();
    cout << solve(0, 0, 0);
    

    return 0;
}
